### PR TITLE
Allow --nodes to accept node names

### PR DESCRIPTION
**Pull request by christarazi**
This flag allows both IP addresses and node names as well. This commit
also adds validation for the IP addresses.

Fixes: https://github.com/cilium/cilium-sysdump/issues/62

### CODE REVIEW

File: cilium-sysdump/__main__.py [hunk 1]
```diff
@@ -23,13 +23,31 @@
...
 log = logging.getLogger(__name__)
 
 
+def is_ip_or_name(s):
+    if not s:
+        return False
+
+    if s[0].isdigit():
+        try:
+            socket.inet_aton(s)
```
Comment (Rolinh) [inline cilium-sysdump/__main__.py:38]: `socket.inet_aton` is IPv4 only. As we need to keep python2 support, please use `socket.inet_pton` with `socket.AF_INET` and `socket.AF_INET6` instead.

File: cilium-sysdump/utils.py [hunk 2]
```diff
@@ -224,3 +230,14 @@ def get_container_names_per_pod(pod_namespace, pod_name):
...
     return time.strftime("%Y%m%d-%H%M%S")
+
+
+def is_node_filter_ip_addr(n):
+    if not n:
+        return False
+
+    for c in n:
+        if not c.isdigit() and c != '.':
+            return False
```
Comment (Rolinh) [inline cilium-sysdump/utils.py:241]: This also leaves out IPv6 support.

**Review by Rolinh — CHANGES_REQUESTED** [review]

File: cilium-sysdump/__main__.py (hunk 1 above)
Comment (christarazi) [reply cilium-sysdump/__main__.py]: Thanks! Fixed.

**Comment by Rolinh** [comment]
@christarazi The detection of IP address from a string is a little convoluted. What about something like this instead?
```python
def is_ipv4(s):
    try:
        socket.inet_pton(socket.AF_INET, s)
        return True
    except (TypeError, socket.error):
        pass
    return False


def is_ipv6(s):
    try:
        socket.inet_pton(socket.AF_INET6, s)
        return True
    except (TypeError, socket.error):
        pass
    return False


def is_ipaddress(s):
    return is_ipv4(s) or is_ipv6(s)


def run():
    s = [
        None,
        "",
        "foo",
        "120.1.1.1",
        "::1",
        "192.168.1.1",
        "::ffff:192.0.2.1",
        "2001:DB8::",
        "2600::",
    ]
    for ip in s:
        print(ip, is_ipaddress(ip))


if __name__ == '__main__':
    run()

```

Output:
```
(None, False)
('', False)
('foo', False)
('120.1.1.1', True)
('::1', True)
('192.168.1.1', True)
('::ffff:192.0.2.1', True)
('2001:DB8::', True)
('2600::', True)
```

**Comment by christarazi** [comment]
@Rolinh Thanks, not sure what I didn't think of that. I've added those functions. However, when parsing / validating the `--nodes` arg, we still need to interpret whether the user intended to pass an IP or a node name, therefore I've left some logic to detect a "likely IP address". I've also left a TODO to do a followup after this PR is merged to properly validate the `--nodes` arg if the user passed in a node that doesn't exist (PR https://github.com/cilium/cilium-sysdump/pull/96 brings the infrastructure for that).

File: cilium-sysdump/__main__.py [hunk 3]
```diff
@@ -24,12 +26,31 @@
...
+def is_ip_or_name(s):
+    if not s:
+        return False
+
+    # We first check if we have any indications that the string is an IPv4 or
+    # an IPv6 address. If so, we'll try to convert the string to an IP address.
+    # If that throws, then we know it's invalid and we'll warn the user.
+    if '.' in s or ':' in s:
+        if not net.is_ipaddress(s):
+            raise argparse.ArgumentTypeError('Invalid IP address')
```
Comment (Rolinh) [inline cilium-sysdump/__main__.py:42]: Just one concern here: node names may contain a `.`. Such node names would be impossible to specify by the user with the current implementation. I wonder if we shouldn't just assume that if we can't read `s` as an IP address, then it's likely a node name (at least, until you implement the TODO down below).

**Review by Rolinh — CHANGES_REQUESTED** [review]

File: cilium-sysdump/__main__.py (hunk 3 above)
Comment (christarazi) [reply cilium-sysdump/__main__.py]: Synced offline. Decided to remove the validation logic and simply assume that if the user enters an IP and it is not an IP, then we assume it's a node name. Also, introduced a log msg to indicate to the user what we are filtering on.

File: cilium-sysdump/utils.py [hunk 4]
```diff
@@ -165,15 +167,26 @@ def get_pods_status_iterator_by_labels(label_selector, host_ip_filter,
...
-    host_ip_filter_cmd = host_ip_filter_cmd.format(
-        "|".join(map(str, host_ip_filter)))
+    if node_filter and net.is_ipaddress(node_filter[0]):
+        log.info("filtering on node IP address ...")
+
+        node_filter_cmd = "kubectl get pods --no-headers " \
+            "-o=custom-columns=NAME:.metadata.name," \
+            "HOSTIP:.status.hostIP --all-namespaces | grep -E \"{}\" | " \
+            "awk '{{print $1}}'"
+    else:
```
Comment (Rolinh) [inline cilium-sysdump/utils.py:177]: I think this else clause should only be reachable if `node_filter` is not empty, right?

File: cilium-sysdump/utils.py [hunk 5]
```diff
@@ -165,15 +167,26 @@ def get_pods_status_iterator_by_labels(label_selector, host_ip_filter,
     # feature in the future. The following filter can be extended by
     # modifying the following kubectl custom-columns and the associated
     # grep command.
-    host_ip_filter_cmd = "kubectl get pods --no-headers " \
-        "-o=custom-columns=NAME:.metadata.name," \
-        "HOSTIP:.status.hostIP --all-namespaces | grep -E \"{}\" | " \
-        "awk '{{print $1}}'"
-    host_ip_filter_cmd = host_ip_filter_cmd.format(
-        "|".join(map(str, host_ip_filter)))
+    if node_filter and net.is_ipaddress(node_filter[0]):
```
Comment (Rolinh) [inline cilium-sysdump/utils.py:170]: Is it fair to assume that users will only provide node names or IP addresses?
Alternatively, we could apply `is_ip_address` on every element of the original list to turn it into 2 lists, one of node names and another one of IP addresses.

File: cilium-sysdump/utils.py (hunk 4 above)
Comment (christarazi) [reply cilium-sysdump/utils.py:177]: Good catch

File: cilium-sysdump/utils.py (hunk 5 above)
Comment (christarazi) [reply cilium-sysdump/utils.py]: Good catch, found a bug if the previous implementation.

Comment (christarazi) [reply cilium-sysdump/utils.py]: Summary of the change is:

introduced `get_pods_filtered` which accepts a custom column for `kubectl get pods` and a list of either node names / IPs to filter against the columns. This function allows us to consolidate the duplicate logic to fetch pods for nodes that match an IP or pods that match a node name, thus allowing us to support a mixture of both for `--nodes`.

File: cilium-sysdump/utils.py [hunk 6]
```diff
@@ -118,7 +120,48 @@ class PodStatus(PodStatus_):
     pass
 
 
-def get_pods_status_iterator_by_labels(label_selector, host_ip_filter,
+def get_pods_filtered(custom_columns, filter, must_exist=True):
+    """Returns a list of pods filtered by the provided filter. The pod list is
+       retrieved by using custom columns.
+
+    Args:
+        cmd - the custom columns to fetch, e.g. pod name, node IP, etc. See
```
Comment (Rolinh) [inline cilium-sysdump/utils.py:128]:
```suggestion
        custom_columns - the custom columns to fetch, e.g. pod name, node IP, etc. See
```

File: cilium-sysdump/utils.py [hunk 7]
```diff
@@ -118,7 +120,48 @@ class PodStatus(PodStatus_):
...
+    Returns:
+        A list of strings representing `kubectl get pods` output.
+    """
+    cc_template = "-o=custom-columns={}"
+    cmd_template = "kubectl get pods --no-headers --all-namespaces {} | " \
+        "grep -E \"{}\" | awk '{{print $1}}'"
+
+    filter_cmd = cmd_template.format(
+        cc_template.format(custom_columns),  # Plugs into `-o=custom-columns=`
+        "|".join(map(str, filter))           # Plugs into `grep -E`
```
Comment (Rolinh) [inline cilium-sysdump/utils.py:144]: Thanks for the comments :+1:

File: cilium-sysdump/utils.py [hunk 8]
```diff
@@ -158,35 +201,47 @@ def get_pods_status_iterator_by_labels(label_selector, host_ip_filter,
...
-        filter_output = subprocess.check_output(
-            host_ip_filter_cmd, shell=True, stderr=subprocess.STDOUT,
+    # Separate out IPs and node names if the input from user is mixed.
+    ip_filter = []
+    name_filter = []
+    for elem in node_filter:
+        if net.is_ipaddress(elem):
+            ip_filter.append(elem)
+        else:
+            name_filter.append(elem)
```
Comment (Rolinh) [inline cilium-sysdump/utils.py:211]: nit: a more pythonic way of achieving the same
```suggestion
    ip_filter = [s for s in node_filter if is_ipaddress(s)]
    name_filter = [s for s in node_filter if not is_ipaddress(s)]
```

**Review by Rolinh — APPROVED** [review]
LGTM, thanks!

**Review by jibi — APPROVED** [review]

**Comment by christarazi** [comment]
Rebased. Merging now.

**Pull request merged**